### PR TITLE
Add CTR conversion for categorical features (with bundling support)

### PR DESCRIPTION
**Pull request by shiyu1994**
This version keeps the original categorical split decisions in trees. We will change to pure numerical decisions in trees soon by recording the mappings from categorical values to CTR values in model. The modification so far only handles the ordinary train and prediction process of LightGBM, other functions like ForceSplits requires further check.

### CODE REVIEW

**Comment by guolinke** [comment]
@shiyu1994 could you resolve the conflict, and provide some benchmark results?

File: include/LightGBM/bin.h [hunk 1]
```diff
@@ -128,6 +128,13 @@ class BinMapper {
   */
   inline uint32_t ValueToBin(double value) const;
 
+  /*!
+  * \brief Whether has value in categorical to bin
+  * \param value
+  * \return whether has value in categorical to bin
+  */
+  inline bool HasValueInCat(double value) const;
```
Comment (guolinke) [inline include/LightGBM/bin.h]: it seems this is not used anymore.

File: include/LightGBM/bin.h [hunk 2]
```diff
@@ -148,7 +155,7 @@ class BinMapper {
   * \param zero_as_missing True to use zero as missing value
   */
   void FindBin(double* values, int num_values, size_t total_sample_cnt, int max_bin, int min_data_in_bin, int min_split_data, BinType bin_type,
-               bool use_missing, bool zero_as_missing);
+               bool use_missing, bool zero_as_missing, bool use_ctr);
```
Comment (guolinke) [inline include/LightGBM/bin.h]: remove `use_ctr`

**Comment by guolinke** [comment]
@shiyu1994 could you rewrite this when have time?

**Comment by shiyu1994** [comment]
> @shiyu1994 could you rewrite this when have time?

Ok. I will rewrite it for the latest version of master branch.

**Comment by shiyu1994** [comment]
Replaced by Pull Request #3234

**Pull request closed without merging**